src/devices/src/legacy/acpi_device.rs

```diff
@@ -0,0 +1,166 @@
+// |           PWRBTN_STS              PWRBTN_EN             |               SLP_TYPx    SLP_TYPx    SLP_TYPx    SLP_EN           | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+
+// To trigger a shutdown, PWRBTN_STS and PWRBTN_EN (in X_PM1a_EVT_BLK) needs to be written as
+// ones and the interrupt should be raised. When guest goes to soft-off state, SLP_EN is written
+// as one and SLP_TYPx is written with the value from the _S5 object from the DSDT table (in
+// fact, when guest goes into power state x, then the value from the _Sx object is written into
+// SLP_TYPx bits located into X_PM1a_CNT_BLK)
+
+pub struct AcpiDevice {
+    /// Data Register
+    data: u64,
+    /// Interrupt to be triggered for OSPM to read ACPI registers
+    interrupt: EventFd,
+    /// CPU reset eventfd. We will set this event to stop Firecracker.
+    reset_evt: EventFd,
+}
+
+impl AcpiDevice {
+    /// Constructs a device that holds ACPI registers.
+    pub fn new(interrupt: EventFd, reset_evt: EventFd) -> Self {
+        Self {
+            data: 0,
+            interrupt,
+            reset_evt,
+        }
+    }
+
+    pub fn send_acpi_shutdown_signal(&mut self) -> Result<()> {
+        const PWRBTN_STS_OFFSET: u64 = 8;
+        const PWRBTN_EN: u64 = 24;
+
+        // Set bits according to the spec in order to trigger the shutdown
+        // https://uefi.org/specs/ACPI/6.4/04_ACPI_Hardware_Specification/ACPI_Hardware_Specification.html#pm1-event-grouping
+        self.data = self.data | (1 << PWRBTN_STS_OFFSET) | (1 << PWRBTN_EN);
+
+        // Make OSPM read ACPI registers and generate the ACPI event by setting the interrupt on ACPI Device
+        self.interrupt
+            .write(9) // SCI_INT value from ACPI FADT table
+            .map_err(Error::AcpiDeviceInterruptFailure)
+    }
+
+    fn shutdown_vmm(&mut self) -> Result<()> {
+        self.reset_evt
+            .write(1)
+            .map_err(Error::AcpiDeviceShutdownFirecrackerFailure)
+    }
+}
+
+impl BusDevice for AcpiDevice {
+    fn read(&mut self, offset: u64, data: &mut [u8]) {
+        const SLP_EN_OFFSET: u64 = 45;
+
+        let mut value;
+
+        if data.len() <= 6 {
+            value = self.data;
+
+            // SLP_EN is a write only bit; reading from it will always return 0 according to the spec:
+            // https://uefi.org/specs/ACPI/6.4/04_ACPI_Hardware_Specification/ACPI_Hardware_Specification.html#pm1-control-registers-fixed-hardware-feature-control-bits
+            value = value & (!(1 << SLP_EN_OFFSET));
+
+            // need to read bytes from given offset; bytes from 0 to offset - 1 should be skipped
+            value = value >> (8 * offset);
+
+            // read each byte from AcpiDevice
+            for i in 0..(data.len()) {
+                data[i] = (value % 256) as u8;
+                value = value >> 8;
+            }
```

Comment:
This looks very specific to x86_64 endianness. While this is fine for current purpose of the PR, I would leave a warning comment for future readers of the code. Smth like
```
// warning: hardcoded little endian logic ahead.
```

(same for `fn write`)

Reply:
I improved this piece of code by using functions like from_le_bytes() and to_le_bytes() that deals with this transformation involving endianness in a more standardized way. Added a comment regarding this, as well.